src/java/org/apache/cassandra/db/compaction/ShardManagerNodeAware.java

```diff
@@ -0,0 +1,312 @@
+            }
+            var splitPoints = findTokenAlignedSplitPoints(sortedTokens.toArray(TOKENS), splitPointCount);
+            logger.debug("Creating shard boundaries for {} shards. Currently {} tokens: {}. Split points: {}", shardCount, sortedTokens.size(), sortedTokens, Arrays.toString(splitPoints));
+            return new NodeAlignedShardTracker(splitPoints);
+        }
+        catch (Throwable t)
+        {
+            logger.error("Error creating shard boundaries", t);
+            throw t;
+        }
+    }
+
+    private Token[] findTokenAlignedSplitPoints(Token[] sortedTokens, int splitPointCount)
+    {
+        assert splitPointCount <= sortedTokens.length : splitPointCount + " > " + sortedTokens.length;
+
+        // Short circuit on equal and on count 1.
+        if (sortedTokens.length == splitPointCount)
+            return sortedTokens;
+        if (splitPointCount == 0)
+            return TOKENS;
+
+        var evenSplitPoints = computeUniformSplitPoints(tokenMetadata.partitioner, splitPointCount);
+        logger.debug("Even split points: {}", Arrays.toString(evenSplitPoints));
+        var nodeAlignedSplitPoints = new Token[splitPointCount];
+
+        // UCS requires that the splitting points for a given density are also splitting points for
+        // all higher densities, so we pick from among the existing tokens.
+        int pos = 0;
+        for (int i = 0; i < evenSplitPoints.length; i++)
+        {
+            Token value = evenSplitPoints[i];
+            pos = Arrays.binarySearch(sortedTokens, pos, sortedTokens.length, value);
+
+            if (pos >= 0)
+            {
+                // Exact match found
+                nodeAlignedSplitPoints[i] = sortedTokens[pos];
+                pos++;
+            }
+            else
+            {
+                // pos is -(insertion point) - 1, so calculate the insertion point
+                pos = -pos - 1;
+
+                if (pos == 0)
+                {
+                    // No left neighbor, so choose the right neighbor
+                    nodeAlignedSplitPoints[i] = sortedTokens[pos];
+                    pos++;
+                }
+                else if (pos == sortedTokens.length)
+                {
+                    // todo assert we're at the end?
+                    // No right neighbor, so choose the left neighbor
+                    nodeAlignedSplitPoints[i] = sortedTokens[pos - 1];
+                    pos++;
+                }
+                else
+                {
+                    // Check the neighbors
+                    Token leftNeighbor = sortedTokens[pos - 1];
+                    Token rightNeighbor = sortedTokens[pos];
+
+                    // Choose the nearest neighbor. By convention, prefer left if value is midpoint, but don't
+                    // choose the same token twice.
+                    if (value.size(leftNeighbor) <= value.size(rightNeighbor) && !leftNeighbor.equals(nodeAlignedSplitPoints[i - 1]))
```

Comment:
Because it's meant to measure the size of potentially wraparound intervals, `size` is not commutative (more precisely, `right.size(left) = 1 - left.size(right)`). It should be called on the left bound with the right as argument.

I.e. this should be `leftNeighbor.size(value) <= value.size(rightNeighbor)`.